<!DOCTYPE html>
<html>

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Habib_2017</title>


<style type="text/css">
body {
  font-family: Helvetica, arial, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  padding-top: 10px;
  padding-bottom: 10px;
  background-color: white;
  padding: 30px; }

body > *:first-child {
  margin-top: 0 !important; }
body > *:last-child {
  margin-bottom: 0 !important; }

a {
  color: #4183C4; }
a.absent {
  color: #cc0000; }
a.anchor {
  display: block;
  padding-left: 30px;
  margin-left: -30px;
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0; }

h1, h2, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
  cursor: text;
  position: relative; }

h1:hover a.anchor, h2:hover a.anchor, h3:hover a.anchor, h4:hover a.anchor, h5:hover a.anchor, h6:hover a.anchor {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA09pVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoMTMuMCAyMDEyMDMwNS5tLjQxNSAyMDEyLzAzLzA1OjIxOjAwOjAwKSAgKE1hY2ludG9zaCkiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OUM2NjlDQjI4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OUM2NjlDQjM4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo5QzY2OUNCMDg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo5QzY2OUNCMTg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PsQhXeAAAABfSURBVHjaYvz//z8DJYCRUgMYQAbAMBQIAvEqkBQWXI6sHqwHiwG70TTBxGaiWwjCTGgOUgJiF1J8wMRAIUA34B4Q76HUBelAfJYSA0CuMIEaRP8wGIkGMA54bgQIMACAmkXJi0hKJQAAAABJRU5ErkJggg==) no-repeat 10px center;
  text-decoration: none; }

h1 tt, h1 code {
  font-size: inherit; }

h2 tt, h2 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h4 tt, h4 code {
  font-size: inherit; }

h5 tt, h5 code {
  font-size: inherit; }

h6 tt, h6 code {
  font-size: inherit; }

h1 {
  font-size: 28px;
  color: black; }

h2 {
  font-size: 24px;
  border-bottom: 1px solid #cccccc;
  color: black; }

h3 {
  font-size: 18px; }

h4 {
  font-size: 16px; }

h5 {
  font-size: 14px; }

h6 {
  color: #777777;
  font-size: 14px; }

p, blockquote, ul, ol, dl, li, table, pre {
  margin: 15px 0; }

hr {
  background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAYAAAAECAYAAACtBE5DAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBNYWNpbnRvc2giIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OENDRjNBN0E2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OENDRjNBN0I2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo4Q0NGM0E3ODY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo4Q0NGM0E3OTY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PqqezsUAAAAfSURBVHjaYmRABcYwBiM2QSA4y4hNEKYDQxAEAAIMAHNGAzhkPOlYAAAAAElFTkSuQmCC) repeat-x 0 0;
  border: 0 none;
  color: #cccccc;
  height: 4px;
  padding: 0;
}

body > h2:first-child {
  margin-top: 0;
  padding-top: 0; }
body > h1:first-child {
  margin-top: 0;
  padding-top: 0; }
  body > h1:first-child + h2 {
    margin-top: 0;
    padding-top: 0; }
body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
  margin-top: 0;
  padding-top: 0; }

a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0; }

h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
  margin-top: 0; }

li p.first {
  display: inline-block; }
li {
  margin: 0; }
ul, ol {
  padding-left: 30px; }

ul :first-child, ol :first-child {
  margin-top: 0; }

dl {
  padding: 0; }
  dl dt {
    font-size: 14px;
    font-weight: bold;
    font-style: italic;
    padding: 0;
    margin: 15px 0 5px; }
    dl dt:first-child {
      padding: 0; }
    dl dt > :first-child {
      margin-top: 0; }
    dl dt > :last-child {
      margin-bottom: 0; }
  dl dd {
    margin: 0 0 15px;
    padding: 0 15px; }
    dl dd > :first-child {
      margin-top: 0; }
    dl dd > :last-child {
      margin-bottom: 0; }

blockquote {
  border-left: 4px solid #dddddd;
  padding: 0 15px;
  color: #777777; }
  blockquote > :first-child {
    margin-top: 0; }
  blockquote > :last-child {
    margin-bottom: 0; }

table {
  padding: 0;border-collapse: collapse; }
  table tr {
    border-top: 1px solid #cccccc;
    background-color: white;
    margin: 0;
    padding: 0; }
    table tr:nth-child(2n) {
      background-color: #f8f8f8; }
    table tr th {
      font-weight: bold;
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr td {
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr th :first-child, table tr td :first-child {
      margin-top: 0; }
    table tr th :last-child, table tr td :last-child {
      margin-bottom: 0; }

img {
  max-width: 100%; }

span.frame {
  display: block;
  overflow: hidden; }
  span.frame > span {
    border: 1px solid #dddddd;
    display: block;
    float: left;
    overflow: hidden;
    margin: 13px 0 0;
    padding: 7px;
    width: auto; }
  span.frame span img {
    display: block;
    float: left; }
  span.frame span span {
    clear: both;
    color: #333333;
    display: block;
    padding: 5px 0 0; }
span.align-center {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-center > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: center; }
  span.align-center span img {
    margin: 0 auto;
    text-align: center; }
span.align-right {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-right > span {
    display: block;
    overflow: hidden;
    margin: 13px 0 0;
    text-align: right; }
  span.align-right span img {
    margin: 0;
    text-align: right; }
span.float-left {
  display: block;
  margin-right: 13px;
  overflow: hidden;
  float: left; }
  span.float-left span {
    margin: 13px 0 0; }
span.float-right {
  display: block;
  margin-left: 13px;
  overflow: hidden;
  float: right; }
  span.float-right > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: right; }

code, tt {
  margin: 0 2px;
  padding: 0 5px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px; }

pre code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent; }

.highlight pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }

pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }
  pre code, pre tt {
    background-color: transparent;
    border: none; }

sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}

kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb
}

* {
	-webkit-print-color-adjust: exact;
}
@media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:0 auto;
    }
}
@media print {
	table, pre {
		page-break-inside: avoid;
	}
	pre {
		word-wrap: break-word;
	}
}
</style>


</head>

<body>

<h2 id="toc_0">Reanalysis of single nucleus RNA-seq (Note S6)</h2>

<p><strong>Step 1. Download snRNA-seq from adult post-mortem human brain tissue</strong>. </p>

<div><pre><code class="language-none">&gt; wget https://storage.googleapis.com/gtex_additional_datasets/single_cell_data/GTEx_droncseq_hip_pcf.tar
&gt; tar xopf GTEx_droncseq_hip_pcf.tar</code></pre></div>

<p><strong>Step 2. Create a snap object</strong></p>

<div><pre><code class="language-R">&gt; R
# Read the Gene Expression Table
&gt; dat = read.table(&quot;GTEx_droncseq_hip_pcf/GTEx_droncseq_hip_pcf.umi_counts.txt.gz&quot;, head=TRUE, row.name=1);
&gt; tsne = read.table(&quot;GTEx_droncseq_hip_pcf/GTEx_droncseq_hip_pcf.tsne.txt.gz&quot;);
&gt; cluster = read.table(&quot;GTEx_droncseq_hip_pcf/GTEx_droncseq_hip_pcf.clusters.txt.gz&quot;)
&gt; cmat = Matrix(t(as.matrix(dat)),sparse=TRUE);
&gt; x.sp = createSnapFromGmat(
                            cmat, 
                            barcodes=rownames(cmat), 
                            geneNames=colnames(cmat)
                            )
&gt; x.sp

number of barcodes: 14963
number of bins: 0
number of peaks: 0
number of genes: 32111
==========================
meta data            (metData) :  FALSE
cellxbin matrix      (bmat)    :  FALSE
cellxpeak matrix     (pmat)    :  FALSE
cellxgene matrix     (gmat)    :  TRUE
jaccard matrix       (jmat)    :  FALSE
normalization        (nmat)    :  FALSE
PCA:                 (smat)    :  FALSE
cluster:             (cluster) :  FALSE
t-sne:               (tsne)    :  FALSE
umap:                (umap)    :  FALSE</code></pre></div>

<p><strong>Step 2. Matrix Binarization</strong></p>

<div><pre><code class="language-R">&gt; x.sp = makeBinary(x.sp, mat=&quot;gmat&quot;);</code></pre></div>

<p><strong>Step 3. Jaccard Matrix &amp; Normlaization</strong></p>

<div><pre><code class="language-none">&gt; x.sp = calJaccard(
        x.sp,
        ncell.chunk=1000,
        mat=&quot;gmat&quot;,
        max.var=5000,
        seed.use=10,
        norm.method=&quot;normOVE&quot;,
        row.center=TRUE,
        row.scale=TRUE,
        low.threshold=-5,
        high.threshold=5,
        keep.jmat=FALSE,
        do.par = TRUE,
        num.cores = 5
        )</code></pre></div>

<p><strong>Step 4. PCA Analysis</strong></p>

<div><pre><code class="language-R">&gt; x.sp = runPCA(
        x.sp,
        pc.num=50,
        input.mat = &quot;nmat&quot;,
        method=&quot;svd&quot;,
        weight.by.sd = FALSE,
        center=TRUE,
        scale=FALSE,
        seed.use=10
        )</code></pre></div>

<p><strong>Step 5. Determine significant principal componentS</strong></p>

<div><pre><code class="language-R"># plot PCA
&gt; plotPCA(x.sp, method=&quot;elbow&quot;);
&gt; plotPCA(x.sp, method=&quot;pairwise&quot;);</code></pre></div>

<p><img src="PCA_elbow_plot.PNG" width="400" height="350" /></p>

<p><img src="PCA_scatter_plot.PNG" width="400" height="400" /></p>

<p><strong>Step 6. Clustering</strong></p>

<div><pre><code class="language-R"># find clusters
&gt; x.sp = runCluster(
    x.sp,
    pca_dims=1:20,
    k=15,
    resolution=1.0,
    method=&quot;louvain&quot;,
    path_to_louvain=&quot;../../../github/snapR/bin/findCommunityLouvain&quot;
    );</code></pre></div>

<p><strong>Step 7. Visualization by t-SNE</strong></p>

<div><pre><code class="language-none">&gt; x.sp = runViz(
    x.sp, 
    pca_dims=1:20, 
    dims=2, 
    method=&quot;Rtsne&quot;
    );
&gt; plotViz(x.sp, method=&quot;tsne&quot;, pch=19, cex=0.1);
&gt; plot(tsne, col=cluster[,2], cex=0.1, main=&quot;Habib 2017&quot;);
&gt; plot(tsne, col=cluster[,2], cex=0.1, type=&quot;n&quot;);
&gt; text(tsne, label=cluster[,2], cex=0.4, col=cluster[,2], &quot;Habib 2017&quot;);</code></pre></div>

<p><img src="Viz_tsne.PNG" width="300" height="300" />
<img src="Viz_tsne_label.PNG" width="300" height="300" />
<img src="Viz_tsne_old.PNG" width="300" height="300" />
<img src="Viz_tsne_old_label.PNG" width="300" height="300" /></p>

<p><strong>Step 8. Marker Gene plot</strong></p>

<div><pre><code class="language-none">&gt; marker.genes = c(&quot;SNAP25&quot;, &quot;GAD2&quot;, &quot;APOE&quot;, &quot;C1QB&quot;, &quot;PVALB&quot;, 
                   &quot;VIP&quot;, &quot;SST&quot;, &quot;LAMP5&quot;, &quot;NPY&quot;, &quot;SLC17A7&quot;, 
                   &quot;MOG&quot;, &quot;PDGFRA&quot;, &quot;CSPG4&quot;,&quot;CX3CR1&quot;,&quot;F3&quot;,&quot;AQP4&quot;
                   );
&gt; plotGene(x.sp, gene.sel= marker.genes, method=&quot;tsne&quot;, p=1, cex=0.1);</code></pre></div>

<p><img src="gene_plot.PNG" width="500" height="500" /></p>

<p><strong>Step 9. Comparision between PFC and HP</strong></p>

<div><pre><code class="language-none">&gt; plot(x.sp@tsne, col=&quot;grey&quot;, cex=0.5, xlab=&quot;tsne-1&quot;, ylab=&quot;tsne2&quot;);
&gt; points(x.sp@tsne[grep(&quot;PFC&quot;, x.sp@barcode),], col=&quot;black&quot;, cex=0.5);
&gt; legend(-40, 45, legend=c(&quot;PFC&quot;, &quot;CA&quot;),
       col=c(&quot;black&quot;, &quot;grey&quot;), lty=1, cex=0.8,
       title=&quot;Line types&quot;, text.font=4, bg=&#39;lightblue&#39;)
&gt; barplot(table(x.sp@cluster[grep(&quot;PFC&quot;, x.sp@barcode)])/table(x.sp@cluster), main=&quot;PFC ratio (expected=0.39)&quot;) </code></pre></div>

<p><img src="PFC_CA.PNG" width="250" height="250" />
<img src="PFC_ratio.PNG" width="250" height="250" /></p>

<p><strong>Step 10. Sub-clustering of GABAergic neurons</strong></p>

<div><pre><code class="language-R"># sub-divison of GABA
&gt; x.gaba.sp = x.sp[which(x.sp@cluster==3),]
&gt; x.gaba.sp = calJaccard(
        x.gaba.sp,
        mat=&quot;gmat&quot;,
        ncell.chunk=1000,
        max.var=5000,
        seed.use=10,
        norm.method=&quot;normOVE&quot;,
        row.center=TRUE,
        row.scale=TRUE,
        low.threshold=-5,
        high.threshold=5,
        keep.jmat=FALSE,
        do.par = TRUE,
        num.cores = 5
        )

&gt; x.gaba.sp = runPCA(
        x.gaba.sp,
        pc.num=50,
        input.mat = &quot;nmat&quot;,
        method=&quot;svd&quot;,
        weight.by.sd = TRUE,
        center=TRUE,
        scale=FALSE,
        seed.use=10
        )

&gt; x.gaba.sp = runCluster(
    x.gaba.sp,
    pca_dims=1:10,
    k=30,
    resolution=1,
    method=&quot;louvain&quot;,
    path_to_louvain=&quot;../../../github/snapR/bin/findCommunityLouvain&quot;
    )

&gt; x.gaba.sp = runViz(
    x.gaba.sp, 
    pca_dims=1:10, 
    dims=2, 
    method=&quot;Rtsne&quot;,
    )

&gt; plotViz(x.gaba.sp, method=&quot;tsne&quot;, pch=19, cex=0.5);
</code></pre></div>

<p><img src="Viz_tsne_GABA.PNG" width="250" height="250" />
<img src="Viz_tsne_GABA_label.PNG" width="250" height="250" /></p>

<div><pre><code class="language-none">&gt; marker.genes = c(&quot;GAD2&quot;, &quot;CCK&quot;, &quot;VIP&quot;, &quot;SST&quot;, &quot;PVALB&quot;, &quot;LAMP5&quot;,&quot;NPY&quot;, &quot;CHODL&quot;, &quot;CDK6&quot;, &quot;MYH8&quot;, &quot;TH&quot;, &quot;SNCG&quot;);
&gt; plotGene(x.gaba.sp, gene.sel= marker.genes, method=&quot;tsne&quot;, p=1, cex=0.6, plot.row=3, plot.col=3);</code></pre></div>

<p><img src="gene_plot_GABA.PNG" width="500" height="500" /></p>




</body>

</html>
