<!DOCTYPE html>
<html>

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Cusanovich_2018</title>


<style type="text/css">
body {
  font-family: Helvetica, arial, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  padding-top: 10px;
  padding-bottom: 10px;
  background-color: white;
  padding: 30px; }

body > *:first-child {
  margin-top: 0 !important; }
body > *:last-child {
  margin-bottom: 0 !important; }

a {
  color: #4183C4; }
a.absent {
  color: #cc0000; }
a.anchor {
  display: block;
  padding-left: 30px;
  margin-left: -30px;
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0; }

h1, h2, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
  cursor: text;
  position: relative; }

h1:hover a.anchor, h2:hover a.anchor, h3:hover a.anchor, h4:hover a.anchor, h5:hover a.anchor, h6:hover a.anchor {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA09pVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoMTMuMCAyMDEyMDMwNS5tLjQxNSAyMDEyLzAzLzA1OjIxOjAwOjAwKSAgKE1hY2ludG9zaCkiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OUM2NjlDQjI4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OUM2NjlDQjM4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo5QzY2OUNCMDg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo5QzY2OUNCMTg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PsQhXeAAAABfSURBVHjaYvz//z8DJYCRUgMYQAbAMBQIAvEqkBQWXI6sHqwHiwG70TTBxGaiWwjCTGgOUgJiF1J8wMRAIUA34B4Q76HUBelAfJYSA0CuMIEaRP8wGIkGMA54bgQIMACAmkXJi0hKJQAAAABJRU5ErkJggg==) no-repeat 10px center;
  text-decoration: none; }

h1 tt, h1 code {
  font-size: inherit; }

h2 tt, h2 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h4 tt, h4 code {
  font-size: inherit; }

h5 tt, h5 code {
  font-size: inherit; }

h6 tt, h6 code {
  font-size: inherit; }

h1 {
  font-size: 28px;
  color: black; }

h2 {
  font-size: 24px;
  border-bottom: 1px solid #cccccc;
  color: black; }

h3 {
  font-size: 18px; }

h4 {
  font-size: 16px; }

h5 {
  font-size: 14px; }

h6 {
  color: #777777;
  font-size: 14px; }

p, blockquote, ul, ol, dl, li, table, pre {
  margin: 15px 0; }

hr {
  background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAYAAAAECAYAAACtBE5DAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBNYWNpbnRvc2giIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OENDRjNBN0E2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OENDRjNBN0I2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo4Q0NGM0E3ODY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo4Q0NGM0E3OTY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PqqezsUAAAAfSURBVHjaYmRABcYwBiM2QSA4y4hNEKYDQxAEAAIMAHNGAzhkPOlYAAAAAElFTkSuQmCC) repeat-x 0 0;
  border: 0 none;
  color: #cccccc;
  height: 4px;
  padding: 0;
}

body > h2:first-child {
  margin-top: 0;
  padding-top: 0; }
body > h1:first-child {
  margin-top: 0;
  padding-top: 0; }
  body > h1:first-child + h2 {
    margin-top: 0;
    padding-top: 0; }
body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
  margin-top: 0;
  padding-top: 0; }

a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0; }

h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
  margin-top: 0; }

li p.first {
  display: inline-block; }
li {
  margin: 0; }
ul, ol {
  padding-left: 30px; }

ul :first-child, ol :first-child {
  margin-top: 0; }

dl {
  padding: 0; }
  dl dt {
    font-size: 14px;
    font-weight: bold;
    font-style: italic;
    padding: 0;
    margin: 15px 0 5px; }
    dl dt:first-child {
      padding: 0; }
    dl dt > :first-child {
      margin-top: 0; }
    dl dt > :last-child {
      margin-bottom: 0; }
  dl dd {
    margin: 0 0 15px;
    padding: 0 15px; }
    dl dd > :first-child {
      margin-top: 0; }
    dl dd > :last-child {
      margin-bottom: 0; }

blockquote {
  border-left: 4px solid #dddddd;
  padding: 0 15px;
  color: #777777; }
  blockquote > :first-child {
    margin-top: 0; }
  blockquote > :last-child {
    margin-bottom: 0; }

table {
  padding: 0;border-collapse: collapse; }
  table tr {
    border-top: 1px solid #cccccc;
    background-color: white;
    margin: 0;
    padding: 0; }
    table tr:nth-child(2n) {
      background-color: #f8f8f8; }
    table tr th {
      font-weight: bold;
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr td {
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr th :first-child, table tr td :first-child {
      margin-top: 0; }
    table tr th :last-child, table tr td :last-child {
      margin-bottom: 0; }

img {
  max-width: 100%; }

span.frame {
  display: block;
  overflow: hidden; }
  span.frame > span {
    border: 1px solid #dddddd;
    display: block;
    float: left;
    overflow: hidden;
    margin: 13px 0 0;
    padding: 7px;
    width: auto; }
  span.frame span img {
    display: block;
    float: left; }
  span.frame span span {
    clear: both;
    color: #333333;
    display: block;
    padding: 5px 0 0; }
span.align-center {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-center > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: center; }
  span.align-center span img {
    margin: 0 auto;
    text-align: center; }
span.align-right {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-right > span {
    display: block;
    overflow: hidden;
    margin: 13px 0 0;
    text-align: right; }
  span.align-right span img {
    margin: 0;
    text-align: right; }
span.float-left {
  display: block;
  margin-right: 13px;
  overflow: hidden;
  float: left; }
  span.float-left span {
    margin: 13px 0 0; }
span.float-right {
  display: block;
  margin-left: 13px;
  overflow: hidden;
  float: right; }
  span.float-right > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: right; }

code, tt {
  margin: 0 2px;
  padding: 0 5px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px; }

pre code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent; }

.highlight pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }

pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }
  pre code, pre tt {
    background-color: transparent;
    border: none; }

sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}

kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb
}

* {
	-webkit-print-color-adjust: exact;
}
@media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:0 auto;
    }
}
@media print {
	table, pre {
		page-break-inside: avoid;
	}
	pre {
		word-wrap: break-word;
	}
}
</style>


</head>

<body>

<h2 id="toc_0">Re-analysis of mouse pre-frontal cortex (PFC) from Cusanovich 2018</h2>

<p><strong>Step 1. Download pre-frontal cortex sciATAC-seq from UW mouse atlas website</strong>. From the bam header, we can tell the reads are mapped to mm9 using bowtie2 with <code>basic-0 -p 12 -X 2000 -3 1 -x</code> and sorted by coordiantes. In order to generate <code>snap</code> file, we first need to resort the reads based on the read name.</p>

<div><pre><code class="language-none">&gt; wget http://krishna.gs.washington.edu/content/members/ajh24/mouse_atlas_data_release/bams/PreFrontalCortex_62216.bam
&gt; samtools view PreFrontalCortex_62216.bam -H
&gt; 
@HD VN:1.0  SO:coordinate
@SQ SN:chr1 LN:197195432
@SQ SN:chr2 LN:181748087
@SQ SN:chr3 LN:159599783
@SQ SN:chr4 LN:155630120
@SQ SN:chr5 LN:152537259
@SQ SN:chr6 LN:149517037
@SQ SN:chr7 LN:152524553
@SQ SN:chr8 LN:131738871
@SQ SN:chr9 LN:124076172
@SQ SN:chr10    LN:129993255
@SQ SN:chr11    LN:121843856
@SQ SN:chr12    LN:121257530
@SQ SN:chr13    LN:120284312
@SQ SN:chr14    LN:125194864
@SQ SN:chr15    LN:103494974
@SQ SN:chr16    LN:98319150
@SQ SN:chr17    LN:95272651
@SQ SN:chr18    LN:90772031
@SQ SN:chr19    LN:61342430
@SQ SN:chrX LN:166650296
@SQ SN:chrY LN:15902555
@SQ SN:chrM LN:16299
@SQ SN:chr1_random  LN:1231697
@SQ SN:chr3_random  LN:41899
@SQ SN:chr4_random  LN:160594
@SQ SN:chr5_random  LN:357350
@SQ SN:chr7_random  LN:362490
@SQ SN:chr8_random  LN:849593
@SQ SN:chr9_random  LN:449403
@SQ SN:chr13_random LN:400311
@SQ SN:chr16_random LN:3994
@SQ SN:chr17_random LN:628739
@SQ SN:chrX_random  LN:1785075
@SQ SN:chrY_random  LN:58682461
@SQ SN:chrUn_random LN:5900358
@PG ID:bowtie2  PN:bowtie2  VN:2.2.3    CL:&quot;/net/gs/vol3/software/modules-sw/bowtie2/2.2.3/Linux/RHEL6/x86_64/bowtie2-align-s --wrapper basic-0 -p 12 -X 2000 -3 1 -x /net/shendure/vol7/cusanovi/genomes/mm9/tophat/mm9 -1 /tmp/43657.inpipe1 -2 /tmp/43657.inpipe2&quot;
@PG ID:bowtie2-23511DC3 PN:bowtie2  VN:2.2.3    CL:&quot;/net/gs/vol3/software/modules-sw/bowtie2/2.2.3/Linux/RHEL6/x86_64/bowtie2-align-s --wrapper basic-0 -p 12 -X 2000 -3 1 -x /net/shendure/vol7/cusanovi/genomes/mm9/tophat/mm9 -1 /tmp/40271.inpipe1 -2 /tmp/40271.inpipe2&quot;

&gt; samtools sort -n -@ 5 PreFrontalCortex_62216.bam -o PreFrontalCortex_62216.nsrt.bam</code></pre></div>

<p><strong>Step 2. Pre-processing (snaptools)</strong>. 
After alignment and sorting, we next create a snap file. </p>

<div><pre><code class="language-bash">&gt; wget http://hgdownload.cse.ucsc.edu/goldenPath/mm9/bigZips/mm9.chrom.sizes
&gt; snaptools snap-pre                                  \
    --input-file=PreFrontalCortex_62216.nsrt.bam      \
    --output-snap=PreFrontalCortex_62216.snap        \
    --genome-name=mm9                                 \
    --genome-size=mm9.chrom.sizes                     \
    --min-mapq=30                                     \
    --min-flen=0                                      \
    --max-flen=1000                                   \
    --keep-chrm=FALSE                                 \
    --keep-single=TRUE                                \
    --keep-secondary=False                            \
    --overwrite=True                                  \
    --min-cov=100                                     \
    --verbose=True</code></pre></div>

<p><strong>Step 3. Cell-by-Bin Matrix Generation (snaptools)</strong>. 
Using generated snap file, we next create the cell-by-bin matrix. Snap file allows for storing cell-by-bin matrices of different resolutions. In the below example, three cell-by-bin matrices are created with bin size of 1,000, 5,000 and 10,000. </p>

<div><pre><code class="language-bash">&gt; snaptools snap-add-bmat                       \
    --snap-file=PreFrontalCortex_62216.snap     \
    --bin-size-list 5000             </code></pre></div>

<p><strong>Step 4. Cell-by-gene Matrix Generation (snaptools)</strong>. 
We next create the cell-by-gene matrix which is later used for cluster annotation.</p>

<div><pre><code class="language-bash">&gt; wget http://renlab.sdsc.edu/r3fang/share/Fang_2019/published_scATAC/Cusanovich_2018/gencode.vM1.annotation.gene.bed;
&gt; snaptools snap-add-gmat                       \
    --snap-file=PreFrontalCortex_62216.snap     \
    --gene-file=gencode.vM1.annotation.gene.bed  </code></pre></div>

<p><strong>Step 5. Create a snap object (snapATAC)</strong></p>

<div><pre><code class="language-R">&gt; R
&gt; library(snapATAC)
&gt; x.sp = createSnap(&quot;PreFrontalCortex_62216.snap&quot;, metaData=TRUE);
&gt; plotBarcode(x.sp);                             
# filter cells based on the following cutoffs
&gt; x.sp = filterCells(x.sp, 
                     subset.names=c(&quot;UMI&quot;),
                     low.thresholds=c(1000),
                     high.thresholds=c(Inf)
                     );
&gt; x.sp 

number of barcodes: 6247
number of bins: 0
number of peaks: 0
number of genes: 0
==========================
meta data            (metData) :  TRUE
cellxbin matrix      (bmat)    :  FALSE
cellxpeak matrix     (pmat)    :  FALSE
cellxgene matrix     (gmat)    :  FALSE
jaccard matrix       (jmat)    :  FALSE
normalization        (nmat)    :  FALSE
PCA:                 (smat)    :  FALSE
cluster:             (cluster) :  FALSE
t-sne:               (tsne)    :  FALSE
umap:                (umap)    :  FALSE</code></pre></div>

<p><img src="barcode_distribution.PNG" width="400" height="400" /></p>

<p><strong>Step 6. Add cell-by-bin matrix</strong></p>

<div><pre><code class="language-R"># show what bin sizes exist in Cusanovich_PreFrontalCortex.snap file
&gt; showBinSizes(&quot;PreFrontalCortex_62216.snap&quot;);
&gt; x.sp = addBmat(x.sp, &quot;PreFrontalCortex_62216.snap&quot;, binSize=5000);
&gt; checkBinSize(x.sp);
[1] 0.9895708</code></pre></div>

<p><strong>Step 8. Make it to binary matrix</strong></p>

<div><pre><code class="language-R">&gt; x.sp = makeBinary(x.sp, mat=&quot;bmat&quot;);</code></pre></div>

<p><strong>Step 9. Feature Selection</strong></p>

<div><pre><code class="language-R"># remove bins that overlap with black list
&gt; system(&quot;wget http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm9-mouse/mm9-blacklist.bed.gz&quot;)
&gt; black.list.mm9 = read.table(&quot;mm9-blacklist.bed.gz&quot;);
&gt; black.list.mm9 = GRanges(
                           black.list.mm9[,1],
                           IRanges(black.list.mm9[,2], black.list.mm9[,3])
                           )
&gt; idy1 = queryHits(findOverlaps(x.sp@feature, black.list.mm9));
# remove bins that are mapped to undesired chromsomes
&gt; idy2 = grep(&quot;random|chrM&quot;, x.sp@feature);
# remove bins that are not covered 
&gt; idy3 = which(colSums(x.sp, mat=&quot;bmat&quot;)==0);
&gt; idy = unique(c(idy1, idy2, idy3));
&gt; x.sp = x.sp[,-idy, mat=&quot;bmat&quot;];
&gt; plotBinCoverage(x.sp);
&gt; x.sp = filterBins(
                    x.sp,
                    low.threshold=-2,
                    high.threshold=2,
                    mat=&quot;bmat&quot;
                    )
&gt; x.sp

number of barcodes: 6247
number of bins: 473113
number of peaks: 0
number of genes: 0
==========================
meta data            (metData) :  TRUE
cellxbin matrix      (bmat)    :  TRUE
cellxpeak matrix     (pmat)    :  FALSE
cellxgene matrix     (gmat)    :  FALSE
jaccard matrix       (jmat)    :  FALSE
normalization        (nmat)    :  FALSE
PCA:                 (smat)    :  FALSE
cluster:             (cluster) :  FALSE
t-sne:               (tsne)    :  FALSE
umap:                (umap)    :  FALSE</code></pre></div>

<p><img src="coverage_hist.PNG" width="300" height="250" /></p>

<p><strong>Step 10. Jaccard Index Matrix &amp; Normalization</strong></p>

<div><pre><code class="language-R"># Calculate Jaccard Index Matrix
&gt; x.sp = calJaccard(
    x.sp,
    mat = &quot;bmat&quot;,
    ncell.chunk=5000,
    max.var=5000,
    seed.use=10,
    norm.method=&quot;normOVE&quot;,
    row.center=TRUE,
    row.scale=TRUE,
    low.threshold=-5,
    high.threshold=5,
    keep.jmat=FALSE,
    do.par=TRUE,
    num.cores=3
    )</code></pre></div>

<p><strong>Step 11. PCA analysis</strong></p>

<div><pre><code class="language-none">&gt; x.sp = runPCA(
    x.sp,
    pc.num=50,
    input.mat=&quot;nmat&quot;,
    method=&quot;svd&quot;,
    weight.by.sd = FALSE,
    center=TRUE,
    scale=FALSE,
    seed.use=10
    )</code></pre></div>

<p><strong>Step 12. Determine the significant principle components</strong></p>

<div><pre><code class="language-none">&gt; plotPCA(x.sp, method=&quot;elbow&quot;);
&gt; plotPCA(x.sp, method=&quot;pairwise&quot;);</code></pre></div>

<p><img src="PCA_elbow.PNG" width="400" height="350" /></p>

<p><img src="PCA_scatter.PNG" width="400" height="400" /></p>

<p><strong>Step 13. Find clusters</strong></p>

<div><pre><code class="language-R">&gt; x.sp = runCluster(
                    x.sp,
                    pca_dims=1:20,
                    k=15,
                    resolution=1.0,
                    method=&quot;jaccard_louvain&quot;,
                    path_to_louvain=&quot;../../../../github/snapR/bin/findCommunityLouvain&quot;
                    )</code></pre></div>

<p><strong>Step 14. Visulization</strong></p>

<div><pre><code class="language-none"># umap
&gt; x.sp = runViz(
                x.sp, 
                pca_dims=1:20, 
                dims=2, 
                method=&quot;umap&quot;
                )
# umap plot
&gt; plotViz(x.sp, method=&quot;umap&quot;, pch=19, cex=0.7);</code></pre></div>

<p><img src="Viz_umap.PNG" width="250" height="250" />
<img src="Viz_umap_label.PNG" width="250" height="250" /></p>

<p><strong>Step 15. Calculate cell-by-gene matrix for marker genes</strong></p>

<div><pre><code class="language-R">&gt; library();
&gt; gene.gr &lt;- import(&quot;gencode.vM1.annotation.gtf.gz&quot;);
&gt; gene.gr &lt;- gene.gr[gene.gr$type == &quot;gene&quot;,];
&gt; marker.genes &lt;- c(&quot;Snap25&quot;, &quot;Gad2&quot;, &quot;Apoe&quot;, 
                    &quot;Pvalb&quot;, &quot;C1qb&quot;, &quot;Mog&quot;,
                    &quot;Vip&quot;,    &quot;Sst&quot;, &quot;Slc17a7&quot;
                    );
&gt; gene.sel.gr &lt;- gene.gr[match(marker.genes, gene.gr$gene_name),];
&gt; x.sp = calCellGeneTable(
                          x.sp, 
                          gene.sel.gr, 
                          mat=&quot;bmat&quot;
                          );
&gt; x.sp = scaleCountMatrix(
                          x.sp, 
                          mat=&quot;gmat&quot;, 
                          cov=rowSums(x.sp, mat=&quot;bmat&quot;), 
                          method=&quot;RPM&quot;
                          );
&gt; plotGene(
           obj=x.sp, 
           gene.sel=marker.genes, 
           method=&quot;tsne&quot;, 
           plot.row=3,
           plot.col=3, 
           cex=0.2,
           binary=FALSE
           );</code></pre></div>

<p><img src="gene_plot.PNG" width="500" height="500" /></p>

<p><strong>Step 16. Subclustering of cluster 9 (Sst+Pv)</strong></p>

<div><pre><code class="language-R">&gt; idx = which(x.sp@cluster == 10);
&gt; x.sub.sp = x.sp[idx,]
&gt; x.sub.sp = calJaccard(
    x.sub.sp,
    mat = &quot;bmat&quot;,
    ncell.chunk=5000,
    max.var=5000,
    seed.use=10,
    norm.method=&quot;normOVE&quot;,
    row.center=TRUE,
    row.scale=TRUE,
    low.threshold=-5,
    high.threshold=5,
    keep.jmat=FALSE,
    do.par=TRUE,
    num.cores=3
    )
&gt; x.sub.sp = runPCA(
    x.sub.sp,
    pc.num=30,
    input.mat=&quot;nmat&quot;,
    method=&quot;svd&quot;,
    weight.by.sd = FALSE,
    center=TRUE,
    scale=FALSE,
    seed.use=10
    )
&gt; x.sub.sp@tsne = x.sub.sp@smat[,c(1,2)]
&gt; marker.genes &lt;- c(&quot;Sst&quot;,&quot;Pvalb&quot;);
&gt; gene.sel.gr &lt;- gene.gr[match(marker.genes, gene.gr$gene_name),];
&gt; x.sub.sp = calCellGeneTable(
                          x.sub.sp, 
                          gene.sel.gr, 
                          mat=&quot;bmat&quot;
                          );
&gt; x.sub.sp = scaleCountMatrix(
                          x.sub.sp, 
                          mat=&quot;gmat&quot;, 
                          cov=rowSums(x.sub.sp, mat=&quot;bmat&quot;), 
                          method=&quot;RPM&quot;
                          );
&gt; plotGene(x.sub.sp, gene.sel= marker.genes, method=&quot;tsne&quot;, p=0.05, cex=1);
</code></pre></div>

<p><img src="Sst_Pvalb.PNG" width="500" height="260" /></p>

<p><strong>Step 17. Subclustering of cluster 6 (Vip+Lamp5)</strong></p>

<div><pre><code class="language-R">&gt; idx = which(x.sp@cluster == 6);
&gt; x.sub.sp = x.sp[idx,]
&gt; x.sub.sp = calJaccard(
    x.sub.sp,
    mat = &quot;bmat&quot;,
    ncell.chunk=5000,
    max.var=5000,
    seed.use=10,
    norm.method=&quot;normOVE&quot;,
    row.center=TRUE,
    row.scale=TRUE,
    low.threshold=-5,
    high.threshold=5,
    keep.jmat=FALSE,
    do.par=TRUE,
    num.cores=3
    )
&gt; x.sub.sp = runPCA(
    x.sub.sp,
    pc.num=30,
    input.mat=&quot;nmat&quot;,
    method=&quot;svd&quot;,
    weight.by.sd = FALSE,
    center=TRUE,
    scale=FALSE,
    seed.use=10
    )
&gt; x.sub.sp@tsne = x.sub.sp@smat[,c(1,2)]
&gt; marker.genes &lt;- c(&quot;Vip&quot;,&quot;Lamp5&quot;);
&gt; gene.sel.gr &lt;- gene.gr[match(marker.genes, gene.gr$gene_name),];
&gt; x.sub.sp = calCellGeneTable(
                          x.sub.sp, 
                          gene.sel.gr, 
                          mat=&quot;bmat&quot;
                          );
&gt; x.sub.sp = scaleCountMatrix(
                          x.sub.sp, 
                          mat=&quot;gmat&quot;, 
                          cov=rowSums(x.sub.sp, mat=&quot;bmat&quot;), 
                          method=&quot;RPM&quot;
                          );
&gt; plotGene(x.sub.sp, gene.sel= marker.genes, method=&quot;tsne&quot;, p=0.1, cex=1);
</code></pre></div>

<p><img src="Vip_Npy.PNG" width="500" height="260" /></p>




</body>

</html>
